core/src/main/java/overflowdb/Graph.java

```diff
@@ -0,0 +1,296 @@
+
+public final class Graph implements AutoCloseable {
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  protected final AtomicLong currentId = new AtomicLong(-1L);
+  protected final NodesList nodes = new NodesList(10000);
+  public final IndexManager indexManager = new IndexManager(this);
+  private final Config config;
+  private boolean closed = false;
+
+  protected final Map<String, NodeFactory> nodeFactoryByLabel;
+  protected final Map<String, EdgeFactory> edgeFactoryByLabel;
+
+  protected final OdbStorage storage;
+  protected final Optional<HeapUsageMonitor> heapUsageMonitor;
+  protected final ReferenceManager referenceManager;
+
+  public static Graph open(Config configuration,
+                           List<NodeFactory<?>> nodeFactories,
+                           List<EdgeFactory<?>> edgeFactories) {
+    Map<String, NodeFactory> nodeFactoryByLabel = new HashMap<>(nodeFactories.size());
+    Map<Integer, NodeFactory> nodeFactoryByLabelId = new HashMap<>(nodeFactories.size());
+    Map<String, EdgeFactory> edgeFactoryByLabel = new HashMap<>(edgeFactories.size());
+    nodeFactories.forEach(factory -> nodeFactoryByLabel.put(factory.forLabel(), factory));
+    nodeFactories.forEach(factory -> nodeFactoryByLabelId.put(factory.forLabelId(), factory));
+    edgeFactories.forEach(factory -> edgeFactoryByLabel.put(factory.forLabel(), factory));
+    return new Graph(configuration, nodeFactoryByLabel, nodeFactoryByLabelId, edgeFactoryByLabel);
+  }
+
+  private Graph(Config config,
+                Map<String, NodeFactory> nodeFactoryByLabel,
+                Map<Integer, NodeFactory> nodeFactoryByLabelId,
+                Map<String, EdgeFactory> edgeFactoryByLabel) {
+    this.config = config;
+    this.nodeFactoryByLabel = nodeFactoryByLabel;
+    this.edgeFactoryByLabel = edgeFactoryByLabel;
+
+    NodeDeserializer nodeDeserializer = new NodeDeserializer(
+        this, nodeFactoryByLabelId, config.isSerializationStatsEnabled());
+    if (config.getStorageLocation().isPresent()) {
+      storage = OdbStorage.createWithSpecificLocation(
+          nodeDeserializer,
+          new File(config.getStorageLocation().get()),
+          config.isSerializationStatsEnabled()
+      );
+      initElementCollections(storage);
+    } else {
+      storage = OdbStorage.createWithTempFile(nodeDeserializer, config.isSerializationStatsEnabled());
+    }
+    referenceManager = new ReferenceManager(storage);
+    heapUsageMonitor = config.isOverflowEnabled() ?
+        Optional.of(new HeapUsageMonitor(config.getHeapPercentageThreshold(), referenceManager)) :
+        Optional.empty();
+  }
+
+  private void initElementCollections(OdbStorage storage) {
+    long start = System.currentTimeMillis();
+    final Set<Map.Entry<Long, byte[]>> serializedNodes = storage.allNodes();
+    logger.info("initializing " + serializedNodes.size() + " nodes from existing storage - this may take some time");
```

Comment:
hmm, this one I'd rather keep on INFO, it only happens once when opening the graph and since it's delaying everything else, it's worth having a prominent message about what's going on